torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1943,8 +1943,21 @@ TensorDomain* TensorDomain::flatten(int64_t start_dim, int64_t end_dim) {
   std::vector<IterDomain*> new_root_domain;
   auto inp_domain = noReductions(getMaybeRFactorDomain());
   new_root_domain.reserve(inp_domain.size());
-  for (auto id : inp_domain) {
-    new_root_domain.push_back(id->cloneWithoutRFactor());
+  for (auto i : c10::irange(inp_domain.size())) {
```

Comment:
Saw the test with flatten in it and realized we don't support expand->flatten, so added that. Please review this section (should be pretty straight forward).

Reply:
Looks like flatten is not checking if the merge is legal, for the following example, I get segfault:
```C++
TEST_F(NVFuserTest, FusionReductionFlatten_CUDA) {
  auto fusion = std::make_unique<Fusion>();
  FusionGuard fg(fusion.get());

  auto tv0 = makeConcreteTensor({2, 3});
  fusion->addInput(tv0);

  auto tv1 = sum(tv0, {1});
  auto tv2 = flatten(tv1, 0, 1);
  fusion->addOutput(tv2);

  // auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  // at::manual_seed(0);
  // auto t0 = at::randn({2, 3}, options);

  // FusionExecutorCache executor_cache(std::move(fusion));
  // auto cg_outputs = executor_cache.runFusionWithInputs({t0});
}
```
This is not your fault, it is already existing in the devel branch. I will push a commit to fix.